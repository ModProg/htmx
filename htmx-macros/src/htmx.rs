use htmx_script::{Script, ToJs};
use manyhow::{ensure, Result};
use proc_macro2::{TokenStream, TokenTree};
use quote::ToTokens;
use rstml::atoms::{CloseTag, OpenTag};
use rstml::node::{
    AttributeValueExpr, KeyedAttribute, KeyedAttributeValue, NodeAttribute, NodeBlock, NodeElement,
    NodeName,
};
use rstml::recoverable::Recoverable;
use syn::spanned::Spanned;
use syn::{parse2, Expr, ExprLit, ExprPath, Lit, LitStr, Stmt};

use crate::special_components::Node;
use crate::*;

pub fn htmx(input: TokenStream) -> Result {
    // https://github.com/rust-lang/rust-analyzer/issues/15572
    // let htmx = match (
    //     proc_macro_crate::crate_name("htmx"),
    //     std::env::var("CARGO_CRATE_NAME").as_deref(),
    // ) { (Ok(FoundCrate::Itself), Ok("htmx")) => quote!(crate),
    //   (Ok(FoundCrate::Name(name)), _) => { let ident = Ident::new(&name,
    //   Span::call_site()); quote!(::#ident) } _ => quote!(::htmx),
    // };
    let htmx = htmx_crate();

    let mut input = input.into_iter().peekable();

    let htmx = match input.peek() {
        Some(TokenTree::Ident(ident)) if ident == "crate" => {
            input.next();
            quote!(crate)
        }
        _ => htmx,
    };

    if input.peek().is_none() {
        return Ok(quote!(#htmx::Html::new()));
    }

    let nodes = rstml::Parser::new(
        rstml::ParserConfig::new()
            .recover_block(true)
            .element_close_use_default_wildcard_ident(false)
            .raw_text_elements(["script"].into())
            .custom_node(),
    )
    // TODO parse_recoverable
    .parse_simple(input.collect::<TokenStream>())?
    .into_iter()
    .map(|n| expand_node(n, &htmx, false))
    .collect::<Result>()?;

    Ok(quote! {
        #use #htmx::{ToHtml, Html, IntoHtmlElements};
        {
            use #htmx::native::*;
            let mut $htmx = Html::new();
            #nodes
            $htmx
        }
    })
}

pub fn expand_node(node: Node, htmx: &TokenStream, child: bool) -> Result {
    Ok(match node {
        Node::Comment(_) => todo!("{}", line!()),
        Node::Doctype(_) => todo!("{}", line!()),
        Node::Fragment(_) => todo!("{}", line!()),
        Node::Element(NodeElement {
            open_tag: OpenTag {
                name, attributes, ..
            },
            children,
            close_tag,
            ..
        }) => {
            let script = name.to_string() == "script";
            let (name, custom) = name_to_struct(name, htmx)?;
            let attributes = attributes
                .into_iter()
                .map(|attribute| match attribute {
                    NodeAttribute::Block(attr) => Ok(quote!(custom_attr(#attr, true))),
                    NodeAttribute::Attribute(KeyedAttribute {
                        key,
                        possible_value,
                    }) => match possible_value {
                        KeyedAttributeValue::Binding(_) => todo!("{}", line!()),
                        KeyedAttributeValue::Value(AttributeValueExpr { value, .. }) => {
                            attribute_key_to_fn(key, value, custom)
                        }
                        KeyedAttributeValue::None => attribute_key_to_fn(key, true, custom),
                    },
                })
                .collect::<Result<Vec<_>>>()?;
            let children = if children.is_empty() {
                quote!()
            } else if script {
                let Some(Node::RawText(script)) = children.first() else {
                    unreachable!("script always raw text")
                };
                let script = script.into_token_stream();
                if let Ok(script) = parse2::<LitStr>(script.clone()) {
                    quote!($node = $node.child(#script);)
                } else if let Ok(block) =
                    parse2::<Recoverable<NodeBlock>>(script.clone()).map(Recoverable::inner)
                {
                    quote!($node = $node.child({#[allow(unused_braces)] #block});)
                } else {
                    let script: Script = parse2(script)?;
                    let script = script.to_java_script();
                    quote!($node = $node.child(#script);)
                }
            } else {
                expand_nodes(children, htmx, true)?
            };
            let main = quote!({let mut $node = #name #(.#attributes)*; #children $node.build()});

            let main = match close_tag {
                Some(CloseTag {
                    name: name @ NodeName::Path(_),
                    ..
                }) if !name.is_wildcard() => {
                    // If close_tag was specified, use it so coloring happens
                    quote!({let _: #name; #main})
                }
                _ => main,
            };
            if child {
                quote!($node = $node.child(&#main);)
            } else {
                quote!(#htmx::ToHtml::write_to_html(&#main, &mut $htmx);)
            }
        }
        Node::Block(_) | Node::Text(_) => {
            if child {
                quote!($node = $node.child(&{#[allow(unused_braces)] #node});)
            } else {
                quote!(#htmx::ToHtml::write_to_html(&{#[allow(unused_braces)] #node}, &mut $htmx);)
            }
        }
        Node::RawText(_) => todo!("{}", line!()),
        Node::Custom(c) => c.expand_node(htmx, child)?,
    })
}

pub fn expand_nodes(nodes: Vec<Node>, htmx: &TokenStream, child: bool) -> Result {
    nodes
        .into_iter()
        .map(|n| expand_node(n, htmx, child))
        .collect()
}

fn name_to_struct(name: NodeName, htmx: &TokenStream) -> Result<(TokenStream, bool)> {
    match name {
        NodeName::Path(path) => Ok((quote!(#path::builder()), false)),
        name @ NodeName::Punctuated(_) => {
            let name = name.to_string();
            ensure!(name.to_ascii_lowercase().chars().all(|c| matches!(c, '-' | '.' | '0'..='9' | '_' | 'a'..='z' | '\u{B7}' | '\u{C0}'..='\u{D6}' | '\u{D8}'..='\u{F6}' | '\u{F8}'..='\u{37D}' | '\u{37F}'..='\u{1FFF}' | '\u{200C}'..='\u{200D}' | '\u{203F}'..='\u{2040}' | '\u{2070}'..='\u{218F}' | '\u{2C00}'..='\u{2FEF}' | '\u{3001}'..='\u{D7FF}' | '\u{F900}'..='\u{FDCF}' | '\u{FDF0}'..='\u{FFFD}' | '\u{10000}'..='\u{EFFFF}')),
         "invalid tag name `{name}`, https://html.spec.whatwg.org/multipage/custom-elements.html#prod-potentialcustomelementname"
        );
            Ok((quote!(#htmx::CustomElement::new_unchecked(#name)), true))
        }
        // This {...}
        NodeName::Block(name) => {
            if let [
                Stmt::Expr(
                    Expr::Lit(ExprLit {
                        lit: Lit::Str(name),
                        ..
                    }),
                    None,
                ),
            ] = &name.stmts[..]
            {
                let name = name.value();
                ensure!(name.to_ascii_lowercase().chars().all(|c| matches!(c, '-' | '.' | '0'..='9' | '_' | 'a'..='z' | '\u{B7}' | '\u{C0}'..='\u{D6}' | '\u{D8}'..='\u{F6}' | '\u{F8}'..='\u{37D}' | '\u{37F}'..='\u{1FFF}' | '\u{200C}'..='\u{200D}' | '\u{203F}'..='\u{2040}' | '\u{2070}'..='\u{218F}' | '\u{2C00}'..='\u{2FEF}' | '\u{3001}'..='\u{D7FF}' | '\u{F900}'..='\u{FDCF}' | '\u{FDF0}'..='\u{FFFD}' | '\u{10000}'..='\u{EFFFF}')),
                 "invalid tag name `{name}`, https://html.spec.whatwg.org/multipage/custom-elements.html#prod-potentialcustomelementname"
                );
                Ok((quote!(#htmx::CustomElement::new_unchecked(#name)), true))
            } else {
                Ok((quote!(#htmx::CustomElement::new(#name)), true))
            }
        }
    }
}

fn attribute_key_to_fn(name: NodeName, value: impl ToTokens, custom: bool) -> Result {
    Ok(match name {
        NodeName::Path(ExprPath { path, .. }) if !custom && path.get_ident().is_some() => {
            quote!(#path(#value))
        }
        NodeName::Path(ExprPath { path, .. })
            if path.segments.first().is_some_and(|hx| hx.ident == "hx") =>
        {
            {
                let sident = path
                    .segments
                    .iter()
                    .map(|i| i.ident.to_string().replace('_', "-"))
                    // hx::swap::oob
                    .collect::<Vec<_>>()
                    .join("-");
                quote_spanned!(path.span()=> custom_attr(#sident, #value))
            }
        }
        // This {...}
        name @ (NodeName::Punctuated(_) | NodeName::Path(_)) => {
            let sname = name.to_string();
            quote_spanned!(name.span()=>  custom_attr(#sname, #value))
        }
        name @ NodeName::Block(_) => quote_spanned!(name.span()=>  custom_attr(#name, #value)),
    })
}
