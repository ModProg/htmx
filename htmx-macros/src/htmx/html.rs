use htmx_script::{Script, ToJs};
use manyhow::{ensure, Error, ErrorMessage, Result};
use proc_macro2::TokenStream;
use quote::ToTokens;
use rstml::atoms::{CloseTag, OpenTag};
use rstml::node::{
    AttributeValueExpr, KeyedAttribute, KeyedAttributeValue, NodeAttribute, NodeBlock, NodeElement,
    NodeFragment, NodeName,
};
use rstml::recoverable::Recoverable;
use syn::spanned::Spanned;
use syn::{parse2, Expr, ExprLit, ExprPath, Lit, LitStr, Stmt};

use super::special_components::{Node, Special};
use super::try_into_iter;
use crate::*;

pub fn html(input: TokenStream) -> Result {
    let nodes = rstml::Parser::new(
        rstml::ParserConfig::new()
            .recover_block(true)
            .element_close_use_default_wildcard_ident(false)
            .custom_node::<Special>()
            .raw_text_elements(["script"].into()),
    )
    // TODO parse_recoverable
    .parse_simple(input)?;

    super::expand_nodes(nodes)
}

impl TryFrom<Node> for super::Node {
    type Error = Error;

    fn try_from(value: Node) -> std::result::Result<Self, Self::Error> {
        match value {
            Node::Comment(comment) => bail!(comment, "html comments are not supported"),
            Node::Doctype(doc_type) => bail!(doc_type, "doc typ is set automatically"),
            Node::Fragment(NodeFragment { tag_open, .. }) => bail!(tag_open, "missing tag name"),
            Node::Element(element) => Ok(super::Node::Element(element.try_into()?)),
            Node::Block(block) => Ok(super::Node::Block(block.into_token_stream())),
            Node::Text(text) => Ok(super::Node::String(text.value)),
            Node::RawText(text) => bail!(
                text.into_token_stream().into_iter().next(),
                "expected `<`, `{{` or `\"`"
            ),
            Node::Custom(special) => special.try_into(),
        }
    }
}

impl TryFrom<NodeElement<Special>> for super::Element {
    type Error = Error;

    fn try_from(value: NodeElement<Special>) -> std::result::Result<Self, Self::Error> {
        let NodeElement {
            open_tag,
            children,
            close_tag,
        } = value;
        Ok(super::Element {
            close_tag: close_tag.and_then(|ct| match ct.name {
                NodeName::Path(p) if !ct.name.is_wildcard() => Some(p.into_token_stream()),
                _ => None,
            }),
            attributes: try_into_iter(open_tag.attributes)?,
            body: if !children.is_empty()
                && matches!(&open_tag.name, NodeName::Path(p) if p.path.is_ident("script"))
            {
                let Some(Node::RawText(script)) = children.first() else {
                    unreachable!("script always raw text")
                };
                let script = script.into_token_stream();
                if let Ok(script) = parse2::<LitStr>(script.clone()) {
                    super::ElementBody::Script(super::ScriptBody::String(script))
                } else if let Ok(block) =
                    parse2::<Recoverable<NodeBlock>>(script.clone()).map(Recoverable::inner)
                {
                    super::ElementBody::Script(super::ScriptBody::Expr(block.into_token_stream()))
                } else {
                    let script: Script = parse2(script)?;
                    let script = script.to_java_script();
                    // quote!(__html.body(#script);)
                    super::ElementBody::Script(super::ScriptBody::Expr(script.into_token_stream()))
                }
            } else {
                super::ElementBody::Children(try_into_iter(children)?)
            },
            open_tag: open_tag.name.try_into()?,
        })
    }
}

fn string_from_block(block: &syn::Block) -> Option<&LitStr> {
    if let [
        Stmt::Expr(
            Expr::Lit(ExprLit {
                lit: Lit::Str(lit), ..
            }),
            None,
        ),
    ] = &block.stmts[..]
    {
        Some(lit)
    } else {
        None
    }
}

impl TryFrom<NodeName> for super::OpenTag {
    type Error = Error;

    fn try_from(value: NodeName) -> std::result::Result<Self, Self::Error> {
        Ok(match value {
            NodeName::Path(path) => super::OpenTag::Path(path.into_token_stream()),
            name @ NodeName::Punctuated(_) => {
                super::OpenTag::from_str(name.to_string(), name.span())?
            }
            NodeName::Block(name) => {
                if let Some(name) = string_from_block(&name) {
                    super::OpenTag::from_str(name.value(), name.span())?
                } else {
                    super::OpenTag::Expr(name.into_token_stream())
                }
            }
        })
    }
}

impl TryFrom<NodeAttribute> for super::Attribute {
    type Error = Error;

    fn try_from(value: NodeAttribute) -> std::result::Result<Self, Self::Error> {
        Ok(match value {
            NodeAttribute::Block(name) => super::Attribute {
                key: if let Some(name) = name.try_block().and_then(string_from_block) {
                    super::AttributeKey::from_str(name.value(), name.span())?
                } else {
                    super::AttributeKey::Expr(name.into_token_stream())
                },
                value: None,
            },
            NodeAttribute::Attribute(attribute) => super::Attribute {
                value: attribute.value().map(ToTokens::into_token_stream),
                key: attribute.key.try_into()?,
            },
        })
    }
}

impl TryFrom<NodeName> for super::AttributeKey {
    type Error = Error;

    fn try_from(value: NodeName) -> std::result::Result<Self, Self::Error> {
        Ok(match value {
            NodeName::Path(p) if p.path.get_ident().is_some() => {
                super::AttributeKey::Fn(p.into_token_stream())
            }
            NodeName::Path(p) if p.path.segments.first().is_some_and(|hx| hx.ident == "hx") => {
                let sident = p
                    .path
                    .segments
                    .iter()
                    .map(|i| i.ident.to_string().replace('_', "-"))
                    // hx::swap::oob
                    .collect::<Vec<_>>()
                    .join("-");
                super::AttributeKey::from_str(sident, p.span())?
            }
            key @ (NodeName::Punctuated(_) | NodeName::Path(_)) => {
                super::AttributeKey::from_str(key.to_string(), key.span())?
            }
            NodeName::Block(block) => {
                if let Some(key) = string_from_block(&block) {
                    super::AttributeKey::from_str(key.value(), key.span())?
                } else {
                    super::AttributeKey::Expr(block.into_token_stream())
                }
            }
        })
    }
}

pub fn expand_node(node: Node) -> Result {
    Ok(match node {
        Node::Comment(_) => todo!("{}", line!()),
        Node::Doctype(_) => todo!("{}", line!()),
        Node::Fragment(_) => todo!("{}", line!()),
        Node::Element(NodeElement {
            open_tag: OpenTag {
                name, attributes, ..
            },
            children,
            close_tag,
            ..
        }) => {
            let script = name.to_string() == "script";
            let (name, node_type) = name_to_struct(name)?;
            let attributes = attributes
                .into_iter()
                .map(|attribute| match attribute {
                    NodeAttribute::Block(attr) => Ok(quote!(custom_attr(#attr, true))),
                    NodeAttribute::Attribute(KeyedAttribute {
                        key,
                        possible_value,
                    }) => match possible_value {
                        KeyedAttributeValue::Binding(_) => todo!("{}", line!()),
                        KeyedAttributeValue::Value(AttributeValueExpr { value, .. }) => {
                            attribute_key_to_fn(key, value, matches!(node_type, NodeType::Custom))
                        }
                        KeyedAttributeValue::None => {
                            attribute_key_to_fn(key, true, matches!(node_type, NodeType::Custom))
                        }
                    },
                })
                .collect::<Result<Vec<_>>>()?;
            let children = if children.is_empty() {
                quote!()
            } else if script {
                // TODO scripts
                let Some(Node::RawText(script)) = children.first() else {
                    unreachable!("script always raw text")
                };
                let script = script.into_token_stream();
                if let Ok(script) = parse2::<LitStr>(script.clone()) {
                    // quote!(__html.body(#script);)
                    quote!(::htmx::ToScript::to_script(&#script, &mut __html);)
                } else if let Ok(block) =
                    parse2::<Recoverable<NodeBlock>>(script.clone()).map(Recoverable::inner)
                {
                    // quote!(__html.body({#[allow(unused_braces)] #block});)
                    quote!(::htmx::ToScript::to_script(&{# [allow(unused_braces)] #block}, &mut __html);)
                } else {
                    let script: Script = parse2(script)?;
                    let script = script.to_java_script();
                    // quote!(__html.body(#script);)
                    quote!(::htmx::ToScript::to_script(&#script, &mut __html);)
                }
            } else {
                expand_nodes(children)?
            };
            let close_arg = if matches!(node_type, NodeType::Component) {
                quote!(&mut __html)
            } else {
                quote!()
            };
            let body = if children.is_empty() {
                quote!(.close(#close_arg))
            } else {
                quote!(.body(::htmx::Fragment(|mut __html: &mut ::htmx::Html| {#children}), #close_arg))
            };
            let main = quote!({{let mut __html = #name #(.#attributes)*; __html}#body;});

            match close_tag {
                Some(CloseTag {
                    name: name @ NodeName::Path(_),
                    ..
                }) if !name.is_wildcard() => {
                    // If close_tag was specified, use it so coloring happens
                    quote!({#name::unused(); #main})
                }
                _ => main,
            }
        }
        Node::Block(_) | Node::Text(_) => {
            quote!(::htmx::IntoHtml::into_html({#[allow(unused_braces)] #node}, &mut __html);)
        }
        Node::RawText(_) => todo!("{}", line!()),
        Node::Custom(c) => c.expand_node()?,
    })
}

pub fn expand_nodes(nodes: Vec<Node>) -> Result {
    nodes.into_iter().map(expand_node).collect()
}

pub fn ensure_tag_name(name: String, span: impl ToTokens) -> Result<String, ErrorMessage> {
    ensure!(
        name.to_ascii_lowercase().chars()
            .all(|c| matches!(c, '-' | '.' | '0'..='9' | '_' | 'a'..='z' | '\u{B7}' | '\u{C0}'..='\u{D6}' | '\u{D8}'..='\u{F6}' | '\u{F8}'..='\u{37D}' | '\u{37F}'..='\u{1FFF}' | '\u{200C}'..='\u{200D}' | '\u{203F}'..='\u{2040}' | '\u{2070}'..='\u{218F}' | '\u{2C00}'..='\u{2FEF}' | '\u{3001}'..='\u{D7FF}' | '\u{F900}'..='\u{FDCF}' | '\u{FDF0}'..='\u{FFFD}' | '\u{10000}'..='\u{EFFFF}')),
        span,
        "invalid tag name `{name}`, https://html.spec.whatwg.org/multipage/custom-elements.html#prod-potentialcustomelementname"
        // TODO similar function but with css error: https://drafts.csswg.org/css-syntax-3/#non-ascii-ident-code-point
    );
    Ok(name)
}

enum NodeType {
    Native,
    Component,
    Custom,
}

fn name_to_struct(name: NodeName) -> Result<(TokenStream, NodeType)> {
    match name {
        NodeName::Path(path)
            if path
                .path
                .get_ident()
                .is_some_and(|i| !i.to_string().contains(char::is_uppercase)) =>
        {
            Ok((quote!(#path::new(&mut __html)), NodeType::Native))
        }
        NodeName::Path(path) => Ok((quote!(#path::new()), NodeType::Component)),
        name @ NodeName::Punctuated(_) => {
            let name = ensure_tag_name(name.to_string(), name)?;
            Ok((
                quote!(::htmx::CustomElement::new_unchecked(&mut __html, #name)),
                NodeType::Custom,
            ))
        }
        // This {...}
        NodeName::Block(name) => {
            if let [
                Stmt::Expr(
                    Expr::Lit(ExprLit {
                        lit: Lit::Str(name),
                        ..
                    }),
                    None,
                ),
            ] = &name.stmts[..]
            {
                let name = ensure_tag_name(name.value(), name)?;
                Ok((
                    quote!(::htmx::CustomElement::new_unchecked(&mut __html, #name)),
                    NodeType::Custom,
                ))
            } else {
                Ok((
                    quote!(::htmx::CustomElement::new(&mut __html, #name)),
                    NodeType::Custom,
                ))
            }
        }
    }
}

fn attribute_key_to_fn(name: NodeName, value: impl ToTokens, custom: bool) -> Result {
    Ok(match name {
        NodeName::Path(ExprPath { path, .. }) if !custom && path.get_ident().is_some() => {
            quote!(#path(#value))
        }
        NodeName::Path(ExprPath { path, .. })
            if path.segments.first().is_some_and(|hx| hx.ident == "hx") =>
        {
            {
                let sident = path
                    .segments
                    .iter()
                    .map(|i| i.ident.to_string().replace('_', "-"))
                    // hx::swap::oob
                    .collect::<Vec<_>>()
                    .join("-");
                quote_spanned!(path.span()=> custom_attr(#sident, #value))
            }
        }
        // This {...}
        name @ (NodeName::Punctuated(_) | NodeName::Path(_)) => {
            let sname = name.to_string();
            quote_spanned!(name.span()=>  custom_attr(#sname, #value))
        }
        name @ NodeName::Block(_) => quote_spanned!(name.span()=>  custom_attr(#name, #value)),
    })
}
