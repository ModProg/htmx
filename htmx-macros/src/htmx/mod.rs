#![allow(unused)]
pub mod html;
mod special_components;

pub mod rusty;

use html_escape::{encode_safe, encode_script};
use manyhow::ensure;
use proc_macro2::{Literal, Span};
use syn::spanned::Spanned;
use syn::LitStr;

use super::*;

// pub fn html(input: TokenStream) -> Result {
//     if input.is_empty() {
//         return Ok(quote!(::htmx::Html::new()));
//     }

//     let mut fork = input.clone().into_iter();

//     let first = fork.next();
//     let second = fork.next();

//     // TODO figure out actual differentiator
//     // probably would be, starts with `<` or starts with `{}` or `""` not
// followed     // by `,`

//     if matches!(input.peek(), Some(TokenTree::Punct(punct)) if
// punct.as_char() == '<') {         html::html(input.collect())
//     } else {
//         rusty::html(input.collect())
//     }
// }

fn try_into_iter<T>(
    input: impl IntoIterator<Item = impl TryInto<T, Error = manyhow::Error>>,
) -> Result<Vec<T>> {
    input.into_iter().map(TryInto::try_into).collect()
}

fn expand_nodes(
    nodes: impl IntoIterator<Item = impl TryInto<Node, Error = manyhow::Error>>,
) -> Result {
    let nodes = nodes
        .into_iter()
        .map(TryInto::try_into)
        .collect::<Result<Vec<Node>>>()?;
    Ok(quote! {
        ::htmx::Fragment(move |mut __html: &mut ::htmx::Html| {
            #[allow(unused_braces)]
            {
                use ::htmx::native::*;
                use ::htmx::IntoHtml as _;
                #(#nodes)*
            };
        })
    })
}

enum Node {
    String(LitStr),
    Block(TokenStream),
    If(If),
    For(For),
    While(While),
    FunctionCall(FunctionCall),
    Element(Element),
}

impl ToTokens for Node {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        match self {
            Node::String(lit) => {
                let value = lit.value();
                let value = encode_safe(&value);
                let mut value = Literal::string(&value);
                value.set_span(lit.span());
                quote!(::htmx::IntoHtml::into_html(::htmx::RawSrc::new(#value), &mut __html);).to_tokens(tokens)
            }
            Node::Block(block) => {
                quote!(::htmx::IntoHtml::into_html({#[allow(unused_braces)] {#block}}, &mut __html);).to_tokens(tokens)
            }
            Node::If(if_) => if_.to_tokens(tokens),
            Node::For(for_) => for_.to_tokens(tokens),
            Node::While(while_) => while_.to_tokens(tokens),
            Node::FunctionCall(call) => call.to_tokens(tokens),
            Node::Element(element) => element.to_tokens(tokens)
        }
    }
}

struct If {
    condition: TokenStream,
    then_branch: Vec<Node>,
    else_branch: ElseBranch,
}

impl ToTokens for If {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let Self {
            condition,
            then_branch,
            else_branch,
        } = self;
        quote! {
            if #condition {
                #(#then_branch)*
            } #else_branch
        }
        .to_tokens(tokens)
    }
}

enum ElseBranch {
    None,
    Else(Vec<Node>),
    ElseIf(Box<If>),
}

impl ToTokens for ElseBranch {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        match self {
            ElseBranch::None => {}
            ElseBranch::Else(nodes) => quote!(else {#(#nodes)*}).to_tokens(tokens),
            ElseBranch::ElseIf(if_) => quote!(else #if_).to_tokens(tokens),
        }
    }
}

struct For {
    pat: TokenStream,
    expr: TokenStream,
    body: Vec<Node>,
}

impl ToTokens for For {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let Self { pat, expr, body } = self;
        quote! {
            for #pat in #expr {
                #(#body)*
            }
        }
        .to_tokens(tokens)
    }
}

struct While {
    expr: TokenStream,
    body: Vec<Node>,
}

impl ToTokens for While {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let Self { expr, body } = self;
        quote! {
            while #expr {
                #(#body)*
            }
        }
        .to_tokens(tokens)
    }
}

struct FunctionCall {
    function: TokenStream,
    args: Vec<TokenStream>,
}

impl ToTokens for FunctionCall {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let Self { function, args } = self;
        quote!(::htmx::IntoHtml::into_html(#function(#(Into::into(#args),)*), &mut __html);)
            .to_tokens(tokens)
    }
}

struct Element {
    open_tag: OpenTag,
    close_tag: Option<TokenStream>,
    attributes: Vec<Attribute>,
    body: ElementBody,
}

impl ToTokens for Element {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let Self {
            open_tag,
            close_tag,
            attributes,
            body,
        } = self;

        let close_tag = close_tag.iter();

        let mut attributes = attributes.clone();

        if !matches!(open_tag, OpenTag::Path(_)) {
            for attribute in &mut attributes {
                if let AttributeKey::Fn(name) = &attribute.key {
                    attribute.key = AttributeKey::from_str(name.to_string(), name.span())
                        .expect("idents should be valid attribute keys")
                }
            }
        };

        quote! {
            {{
                #( use ::htmx::__private::Unused; #close_tag::unused(); )*
                #open_tag #(#attributes)* #body
            }.into_html(&mut __html)}
        }
        .to_tokens(tokens)
    }
}

enum OpenTag {
    Path(TokenStream),
    String(String, Span),
    Expr(TokenStream),
}

impl OpenTag {
    fn from_str(name: String, span: Span) -> Result<OpenTag> {
        ensure!(
            name.to_ascii_lowercase().chars()
                .all(|c| matches!(c, '-' | '.' | '0'..='9' | '_' | 'a'..='z' | '\u{B7}' | '\u{C0}'..='\u{D6}' | '\u{D8}'..='\u{F6}' | '\u{F8}'..='\u{37D}' | '\u{37F}'..='\u{1FFF}' | '\u{200C}'..='\u{200D}' | '\u{203F}'..='\u{2040}' | '\u{2070}'..='\u{218F}' | '\u{2C00}'..='\u{2FEF}' | '\u{3001}'..='\u{D7FF}' | '\u{F900}'..='\u{FDCF}' | '\u{FDF0}'..='\u{FFFD}' | '\u{10000}'..='\u{EFFFF}')),
            span,
            "invalid tag name `{name}`, https://html.spec.whatwg.org/multipage/custom-elements.html#prod-potentialcustomelementname"
            // TODO similar function but with css error: https://drafts.csswg.org/css-syntax-3/#non-ascii-ident-code-point
        );
        Ok(OpenTag::String(name, span))
    }
}

impl ToTokens for OpenTag {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        match self {
            OpenTag::Path(path) => quote!(#path::new(&mut __html)),
            OpenTag::String(name, span) => {
                let name = quote_spanned!(*span=> #name);
                quote!(::htmx::CustomElement::new_unchecked(&mut __html, #name))
            }
            OpenTag::Expr(name) => quote!(quote!(::htmx::CustomElement::new(&mut __html, #name)),),
        }
        .to_tokens(tokens)
    }
}

#[derive(Clone)]
struct Attribute {
    key: AttributeKey,
    // TODO value encoding
    value: Option<TokenStream>,
}

impl ToTokens for Attribute {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let Self { key, value } = self;
        let value = value.clone().unwrap_or_else(|| quote!(true));
        match key {
            AttributeKey::Fn(fun) => quote!(.#fun(#value)),
            AttributeKey::String(key, span) => {
                let key = quote_spanned!(*span => #key);
                quote!(.custom_attr_unchecked(#key, #value))
            }
            AttributeKey::Expr(key) => quote!(.custom_attr(#key, #value)),
        }
        .to_tokens(tokens);
    }
}

#[derive(Clone)]
enum AttributeKey {
    Fn(TokenStream),
    String(String, Span),
    Expr(TokenStream),
}

impl AttributeKey {
    fn from_str(value: String, span: Span) -> Result<AttributeKey> {
        ensure!(
            !value.to_string().chars().any(|c| c.is_whitespace()
                || c.is_control()
                || matches!(c, '\0' | '"' | '\'' | '>' | '/' | '=')),
            span,
            "invalid key `{value}`, https://www.w3.org/TR/2011/WD-html5-20110525/syntax.html#attributes-0"
        );
        Ok(AttributeKey::String(value, span))
    }
}

enum ElementBody {
    Script(ScriptBody),
    Children(Vec<Node>),
}

impl ToTokens for ElementBody {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        match self {
            ElementBody::Script(script) => quote!(.body(#script)),
            ElementBody::Children(children) if children.is_empty() => {
                quote!(.close())
            }
            ElementBody::Children(children) => {
                quote!(.body(::htmx::Fragment(|mut __html: &mut ::htmx::Html| {#(#children)*})))
            }
        }
        .to_tokens(tokens)
    }
}

enum ScriptBody {
    String(LitStr),
    Expr(TokenStream),
}

impl ToTokens for ScriptBody {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        match self {
            ScriptBody::String(lit) => {
                let value = lit.value();
                let value = encode_script(&value);
                let mut value = Literal::string(&value);
                value.set_span(lit.span());
                quote!(RawSrc(#value)).to_tokens(tokens)
            }
            ScriptBody::Expr(expr) => expr.to_tokens(tokens),
        }
    }
}
