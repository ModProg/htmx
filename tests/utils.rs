use htmx::{html, HtmlPage};
use insta::assert_snapshot;

#[test]
fn html_page() {
    assert_snapshot!(
        html! {
            <HtmlPage mobile title="Title" lang="de" style_sheets=["hello.css", "world.css"]
                scripts=vec!["a_script.js".to_string()] >
                <a> </a>
            </_>
        }
        .into_string()
        .as_str()
    )
}
