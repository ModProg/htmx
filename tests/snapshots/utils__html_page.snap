---
source: tests/utils.rs
expression: "html! {\n                <HtmlPage mobile title=\"Title\" lang=\"de\"\n                style_sheets=[\"hello.css\", \"world.css\"]\n                scripts=vec![\"a_script.js\".to_string()] > <a> </a> </_>\n            }.to_string().as_str()"
---
<!DOCTYPE html><html lang="de"><head><meta charset="utf-8"><title>Title</title><meta name="viewport" content="width=device-width, initial-scale=1"><link href="hello.css" rel="stylesheet"><link href="world.css" rel="stylesheet"><script src="a_script.js"></script></head><body><a></a></body></html>
